### ISSUE TITLE
torch.Tensor(dtype = torch.uint8) is not OK for index_select?

### ISSUE THREAD

**Issue by huanglin6385** [pytorch/pytorch#17346]
## To Reproduce

import torch
number = torch.randn(6,3,11)
p = torch.arange(10)
print(0,0,p[0],number[0,0,p[0]])
p = torch.arange(10,dtype = torch.uint8)
print(0,0,p[0],number[0,0,p[0]])
print(0,0,p[0],number[0,0,int(p[0])])

output:
0 0 tensor(0) **tensor(-0.2130)**
0 0 tensor(0, dtype=torch.uint8) **tensor([], size=(0, 11))**
0 0 tensor(0, dtype=torch.uint8) **tensor(-0.2130)**

torch.uint8 is not OK for index！

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by SsnL** [comment]
uint8 is used for masking. int64 is for indexing.
